File: ios-native/Sources/CoreDatastore/ServerStore.swift
import Foundation
import Observation

/// Which deployment the app talks to.
///
/// ``MindlogEnvironment/current`` only sets the default; the value can be
/// changed at runtime, which is how a debug build gets pointed at qualif
/// without a recompile. Same arrangement as the Android client's `ServerStore`,
/// with `UserDefaults` where it has DataStore.
@MainActor
@Observable
public final class ServerStore {

    private let defaults: UserDefaults
    public let environment: MindlogEnvironment

    public private(set) var baseURL: URL

    public init(
        defaults: UserDefaults = .standard,
        environment: MindlogEnvironment = .current
    ) {
        self.defaults = defaults
        self.environment = environment
        self.baseURL = defaults.string(forKey: Self.key)
            .flatMap(URL.init(string:))
            ?? environment.defaultBaseURL
    }

    /// The trailing slash is not cosmetic: prod and qualif URLs carry the
    /// `/app` sub-path, and a base without one turns every relative path
    /// resolution into a guess. ``APIURL`` appends rather than resolves, so it
    /// survives either form — but the value is also concatenated as a string to
    /// build the mindlog id URL, and that one does not.
    public func set(_ url: URL) {
        let normalised = url.absoluteString.hasSuffix("/")
            ? url
            : URL(string: url.absoluteString + "/") ?? url
        baseURL = normalised
        defaults.set(normalised.absoluteString, forKey: Self.key)
    }

    public func reset() {
        defaults.removeObject(forKey: Self.key)
        baseURL = environment.defaultBaseURL
    }

    private static let key = "base_url"
}

File: android-native/core/datastore/src/main/kotlin/today/mindlog/todo/core/datastore/ServerStore.kt
package today.mindlog.todo.core.datastore

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import javax.inject.Inject
import javax.inject.Singleton

private val Context.serverDataStore: DataStore<Preferences> by preferencesDataStore(
    name = "mindlog_todo_server",
)

/**
 * Which deployment the app talks to.
 *
 * `-PmindlogEnv` only sets the *default* (see [BuildConfig.DEFAULT_BASE_URL]);
 * the value can be changed at runtime, which is how a debug build gets pointed
 * at qualif without a recompile. Same arrangement as the archived talk client.
 */
@Singleton
class ServerStore @Inject constructor(
    @param:ApplicationContext private val context: Context,
) {
    val baseUrl: Flow<String> = context.serverDataStore.data.map { prefs ->
        prefs[BASE_URL] ?: BuildConfig.DEFAULT_BASE_URL
    }

    suspend fun current(): String = baseUrl.first()

    /**
     * The trailing slash is not cosmetic: prod and qualif URLs carry the `/app`
     * sub-path, and Retrofit drops the last segment of a base URL that lacks
     * one — turning `…/app` into `…/`, which 404s in production while local
     * (no sub-path) keeps working.
     */
    suspend fun set(url: String) {
        val normalised = if (url.endsWith("/")) url else "$url/"
        context.serverDataStore.edit { it[BASE_URL] = normalised }
    }

    suspend fun reset() {
        context.serverDataStore.edit { it.remove(BASE_URL) }
    }

    val environment: String get() = BuildConfig.MINDLOG_ENV

    private companion object {
        val BASE_URL = stringPreferencesKey("base_url")
    }
}
